### PR TITLE
TestRail Client V3

### PR DESCRIPTION
**Pull request by fushinoryuu**
I've been working on this new version for a while, and you can reference some of the conversations on pull request #33.

Here are the changes:

- Sending a request will now return a `RequestResult<T>` object, the object includes:
  - The http status code
  - The payload object of whatever `T` was
  - Any exceptions that might've been thrown
  - And the raw JSON in case the user wants to look at it
  - Returning the `T` object will mirror how TestRail actually behaves instead of returning a ulong like before
- Created a new `TestRailRequest` class to make it easier to build and send requests
- Created enums to make it easier to identify what data is needed in requests instead of using constants everywhere
- Added intellisense documentation where it was missing
- General cleanup and refactoring

I was planning on tackling the other open issues in this pull request, but decided that I didn't want to keep making this pull request any bigger.

Let me know if there are any concerns, thanks!

PS. Once this is merged in, the branch called `v3` can be deleted.

### CODE REVIEW

File: TestRail/TestRailClient.cs [hunk 1]
```diff
@@ -1,1156 +1,1154 @@
...
 
-        /// <summary>Delete the suite</summary>
-        /// <param name="suiteID">id of the suite to delete</param>
-        /// <returns>result of the deletion</returns>
-        public CommandResult<ulong> DeleteSuite(ulong suiteID)
+        /// <summary>Deletes an existing test suite. This action requires the account to have permissions to delete.</summary>
+        /// <param name="suiteId">The ID of the test suite.</param>
+        /// <returns>
+        /// Please note: Deleting a test suite cannot be undone and also deletes all active test runs and results, i.e. test runs and results that weren't closed (archived) yet.
+
```
Comment (Astro03) [inline TestRail/TestRailClient.cs]: I haven't coded C# in years but I think the convention we were doing here was private methods got a "_" prefixed to the name.

Comment (Astro03) [inline TestRail/TestRailClient.cs]: Is there a better way to do this than this if block with multiple returns?

File: TestRail/Types/Case.cs [hunk 2]
```diff
@@ -1,32 +1,33 @@
...
 namespace TestRail.Types
 {
+    /// <inheritdoc />
     /// <summary>stores information about a case</summary>
     public class Case : BaseTestRailType
     {
         #region Public Properties
         /// <summary>id of the case</summary>
-        public ulong? ID { get; set; }
+        public ulong? Id { get; set; }
```
Comment (Astro03) [inline TestRail/Types/Case.cs]: thx for changing these

File: TestRail/Types/Case.cs [hunk 3]
```diff
@@ -57,42 +58,62 @@ public override string ToString()
...
         public virtual JObject GetJson()
         {
             dynamic jsonParams = new JObject();
-            if (!string.IsNullOrWhiteSpace(Title)) { jsonParams.title = Title; }
-            if (null != TypeID) { jsonParams.type_id = TypeID.Value; }
-            if (null != PriorityID) { jsonParams.priority_id = PriorityID.Value; }
-            if (!string.IsNullOrWhiteSpace(Estimate)) { jsonParams.estimate = Estimate; }
-            if (null != MilestoneID) { jsonParams.milestone_id = MilestoneID.Value; }
-            if (!string.IsNullOrWhiteSpace(References)) { jsonParams.refs = References; }
+
```
Comment (Astro03) [inline TestRail/Types/Case.cs]: nit: i feel like this looked cleaner before. IMHO (but i'm not sure with C# convention anymore.

File: TestRail/Types/Milestone.cs [hunk 4]
```diff
@@ -55,38 +57,60 @@ public override string ToString()
...
         /// <returns>json object that represents this class</returns>
         public JObject GetJson()
         {
             dynamic jsonParams = new JObject();
-            if (!string.IsNullOrWhiteSpace(Name)) { jsonParams.name = Name; }
-            if (!string.IsNullOrWhiteSpace(Description)) { jsonParams.description = Description; }
-            if (null != ParentID) { jsonParams.parent_id = ParentID.Value; }
-            if (null != DueOn) { jsonParams.dueOn = DueOn.Value.ToUnixTimestamp(); }
-            if (null != IsCompleted) { jsonParams.is_completed = IsCompleted; }
+
```
Comment (Astro03) [inline TestRail/Types/Milestone.cs]: nit: i feel like it looks cleaner before

File: TestRail/Types/Plan.cs [hunk 5]
```diff
@@ -111,43 +113,61 @@ public static Plan Parse(JObject json)
...
         /// <summary>Get the json object that describes this object</summary>
         /// <returns>the json object</returns>
         public JObject GetJson()
         {
             dynamic jsonParams = new JObject();
-            if (!string.IsNullOrWhiteSpace(Name)) { jsonParams.name = Name; }
-            if (!string.IsNullOrWhiteSpace(Description)) { jsonParams.description = Description; }
-            if (null != MilestoneID) { jsonParams.milestone_id = MilestoneID; }
+
+            if (!string.IsNullOrWhiteSpace(Name))
```
Comment (Astro03) [inline TestRail/Types/Plan.cs]: same

File: TestRail/Types/PlanEntry.cs [hunk 6]
```diff
@@ -42,46 +44,62 @@ public class PlanEntry : BaseTestRailType
...
+            {
+                jsonParams.assignedto_id = AssignedToId.Value;
+            }
+
+            if (null != CaseIds && 0 < CaseIds.Count)
             {
                 var jarray = new JArray();
-                foreach (var caseID in CaseIDs)
+
+                foreach (var caseId in CaseIds)
```
Comment (Astro03) [inline TestRail/Types/PlanEntry.cs]: maybe want to convert these to linq but that's for another PR.

File: TestRail/Utils/RequestResult.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                RawJson = rawJson;
+                Payload = JsonConvert.DeserializeObject<T>(RawJson);
+            }
+
+            StatusCode = status;
+            ThrownException = thrownException;
+        }
+    }
+}
```
Comment (Astro03) [inline TestRail/Utils/RequestResult.cs]: nit newline

File: TestRail/Utils/TestRailRequest.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+            using (var reader = new StreamReader(response.GetResponseStream() ?? throw new InvalidOperationException(), Encoding.UTF8))
+            {
+                var responseFromServer = reader.ReadToEnd();
+
+                return new RequestResult<T>(response.StatusCode, responseFromServer);
+            }
+        }
+    }
+}
```
Comment (Astro03) [inline TestRail/Utils/TestRailRequest.cs]: nit: newline

**Review by Astro03 — APPROVED** [review]

**Comment by fushinoryuu** [comment]
@Astro03 & @Aemilivs: Sorry for taking so long, I had a lot going on and just now got free time to come back to this PR.

I was going to try to write unit tests, but I don't have access to a TestRail instance to test my changes.

**Pull request merged**